File: Piyo/Repositories/Git.swift
import Foundation
import SwiftGitX
import libgit2

/// Git access for the sidebar. Opening a repo and reading its working directory
/// and HEAD branch go through SwiftGitX's Swift API; the two things SwiftGitX
/// has no API for — discovering a repo's toplevel from a subdirectory, and
/// enumerating *linked* worktrees — drop to libgit2's C API (`import libgit2`),
/// the same single libgit2 module SwiftGitX itself builds on.
enum Git {
    /// libgit2 must be initialized once before the raw C calls below. One-shot;
    /// process exit handles the matching shutdown.
    private static let initialized: Void = { git_libgit2_init() }()

    /// Resolve any folder inside a repository to the repo's working-directory
    /// root. SwiftGitX only opens an exact repo path (no discovery), so this uses
    /// libgit2's `git_repository_open_ext`. Returns nil when `folder` isn't part
    /// of a git repository.
    static func resolveToplevel(_ folder: String) -> String? {
        _ = initialized
        var repo: OpaquePointer?
        guard git_repository_open_ext(&repo, folder, 0, nil) == 0 else { return nil }
        defer { git_repository_free(repo) }
        guard let wd = git_repository_workdir(repo) else { return nil }
        return normalize(String(cString: wd))
    }

    /// Every worktree of the repo at `repoPath` — the main worktree first, then
    /// each linked worktree. Branches come from SwiftGitX; the linked-worktree
    /// list comes from libgit2. Returns whatever succeeds; a broken repo yields
    /// an empty list rather than throwing.
    static func worktrees(at repoPath: String) -> [Worktree] {
        _ = initialized
        var seen = Set<String>()
        var result: [Worktree] = []
        func add(_ path: String, _ branch: String) {
            let p = normalize(path)
            if seen.insert(p).inserted { result.append(Worktree(path: p, branch: branch)) }
        }

        // Main worktree (SwiftGitX gives both its path and branch).
        if let main = inspect(repoPath) { add(main.path, main.branch) }
        // Linked worktrees: libgit2 enumerates them, SwiftGitX reads each branch.
        for path in linkedWorktreePaths(repoPath) {
            add(path, inspect(path)?.branch ?? "(detached)")
        }
        return result
    }

    /// Open the repo at `path` with SwiftGitX and read its working directory and
    /// current branch short name. nil if it can't be opened. Works for both a
    /// main repo and a linked worktree (each has its own openable working dir).
    private static func inspect(_ path: String) -> (path: String, branch: String)? {
        guard let repo = try? Repository.open(at: URL(fileURLWithPath: path)),
            let workdir = try? repo.workingDirectory
        else { return nil }

        let branch: String
        if repo.isHEADUnborn {
            branch = "(unborn)"
        } else if repo.isHEADDetached {
            branch = "(detached)"
        } else if let head = try? repo.HEAD {
            branch = head.name  // short name, e.g. "main", via Reference.name
        } else {
            branch = "(detached)"
        }
        return (workdir.path, branch)
    }

    /// Paths of the repo's *linked* worktrees (libgit2's `git_worktree_list`
    /// omits the main one, which `worktrees(at:)` handles separately).
    private static func linkedWorktreePaths(_ repoPath: String) -> [String] {
        var repo: OpaquePointer?
        guard git_repository_open(&repo, repoPath) == 0 else { return [] }
        defer { git_repository_free(repo) }

        var names = git_strarray()
        guard git_worktree_list(&names, repo) == 0, let strings = names.strings else { return [] }
        defer { git_strarray_dispose(&names) }

        var paths: [String] = []
        for i in 0..<Int(names.count) {
            guard let cName = strings[i] else { continue }
            var wt: OpaquePointer?
            guard git_worktree_lookup(&wt, repo, cName) == 0 else { continue }
            defer { git_worktree_free(wt) }
            if let cPath = git_worktree_path(wt) { paths.append(String(cString: cPath)) }
        }
        return paths
    }

    /// Drop libgit2's trailing slash on working-directory paths so paths compare
    /// and display cleanly.
    private static func normalize(_ path: String) -> String {
        var p = path
        while p.count > 1, p.hasSuffix("/") { p.removeLast() }
        return p
    }
}

File: crates/core/src/git.rs
//! Git access for the sidebar, backed by gitoxide (`gix`) — a pure-Rust git
//! implementation (no libgit2). These are internal helpers used by the store
//! ([`crate::store`]); only `Worktree` crosses the UniFFI boundary.

use std::collections::HashSet;
use std::path::Path;

/// One worktree of a git repository: its working-directory path and the short
/// name of the branch checked out there (`(detached)`/`(unborn)` when HEAD
/// isn't a normal branch).
#[derive(uniffi::Record, Clone, Debug)]
pub struct Worktree {
    pub path: String,
    pub branch: String,
}

/// Resolve any folder inside a repository to the repo's working-directory root.
/// Returns `None` when `folder` isn't part of a (non-bare) git repository.
pub(crate) fn resolve_toplevel(folder: &str) -> Option<String> {
    let repo = gix::discover(folder).ok()?;
    repo.workdir().map(normalize)
}

/// Every worktree of the repo at `repo_path` — the main worktree first, then
/// each linked worktree — each with the short name of its checked-out branch.
/// A broken/unopenable repo yields an empty list rather than erroring.
pub(crate) fn worktrees(repo_path: &str) -> Vec<Worktree> {
    let mut seen = HashSet::new();
    let mut out = Vec::new();
    let Ok(repo) = gix::open(repo_path) else {
        return out;
    };

    // Main worktree (its branch read from HEAD).
    if let Some(wd) = repo.workdir() {
        let path = normalize(wd);
        if seen.insert(path.clone()) {
            out.push(Worktree {
                branch: branch_label(&repo),
                path,
            });
        }
    }

    // Linked worktrees: list them, then open each to read its branch.
    if let Ok(proxies) = repo.worktrees() {
        for proxy in proxies {
            let Ok(base) = proxy.base() else { continue };
            let path = normalize(&base);
            if !seen.insert(path.clone()) {
                continue;
            }
            let branch = proxy
                .into_repo()
                .ok()
                .map(|r| branch_label(&r))
                .unwrap_or_else(|| "(detached)".to_string());
            out.push(Worktree { branch, path });
        }
    }
    out
}

/// Short branch name for a repo's HEAD, or `(unborn)`/`(detached)` when HEAD
/// isn't pointing at an existing branch.
fn branch_label(repo: &gix::Repository) -> String {
    match repo.head() {
        Ok(head) => match head.kind {
            gix::head::Kind::Symbolic(reference) => reference.name.shorten().to_string(),
            gix::head::Kind::Unborn(_) => "(unborn)".to_string(),
            gix::head::Kind::Detached { .. } => "(detached)".to_string(),
        },
        Err(_) => "(detached)".to_string(),
    }
}

/// Drop a trailing slash so paths compare and display cleanly.
fn normalize(p: &Path) -> String {
    let s = p.to_string_lossy();
    let trimmed = s.trim_end_matches('/');
    if trimmed.is_empty() {
        "/".to_string()
    } else {
        trimmed.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;
    use std::path::Path;
    use std::process::Command;

    fn git(dir: &Path, args: &[&str]) {
        let status = Command::new("git")
            .current_dir(dir)
            .args(args)
            .env("GIT_AUTHOR_NAME", "t")
            .env("GIT_AUTHOR_EMAIL", "t@t")
            .env("GIT_COMMITTER_NAME", "t")
            .env("GIT_COMMITTER_EMAIL", "t@t")
            .status()
            .unwrap();
        assert!(status.success(), "git {args:?} failed");
    }

    #[test]
    fn resolves_toplevel_from_subdir() {
        // This crate lives in a git repo; `src/` is a subdirectory of its root.
        let sub = format!("{}/src", env!("CARGO_MANIFEST_DIR"));
        let top = resolve_toplevel(&sub).expect("should discover the repo");
        assert!(top.ends_with("/piyo"), "unexpected toplevel: {top}");
    }

    #[test]
    fn not_a_repo_is_none() {
        assert_eq!(resolve_toplevel("/"), None);
    }

    #[test]
    fn lists_main_and_linked_worktrees() {
        let tmp = std::env::temp_dir().join(format!("piyo-gixtest-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&tmp);
        let main = tmp.join("main");
        std::fs::create_dir_all(&main).unwrap();

        git(&main, &["-c", "init.defaultBranch=main", "init"]);
        std::fs::write(main.join("f.txt"), "hi").unwrap();
        git(&main, &["add", "."]);
        git(&main, &["commit", "-m", "init"]);
        // A linked worktree on a new branch `feature`, as a sibling dir.
        git(&main, &["worktree", "add", "-b", "feature", "../wt"]);

        let wts = worktrees(&main.to_string_lossy());
        let branches: HashSet<_> = wts.iter().map(|w| w.branch.as_str()).collect();
        assert_eq!(wts.len(), 2, "got: {wts:?}");
        assert!(branches.contains("main"), "got: {branches:?}");
        assert!(branches.contains("feature"), "got: {branches:?}");

        let _ = std::fs::remove_dir_all(&tmp);
    }

    #[test]
    fn unborn_head_is_labeled() {
        let dir = std::env::temp_dir().join(format!("piyo-gixunborn-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        git(&dir, &["-c", "init.defaultBranch=main", "init"]);

        let wts = worktrees(&dir.to_string_lossy());
        assert_eq!(wts.len(), 1, "got: {wts:?}");
        assert_eq!(wts[0].branch, "(unborn)");

        let _ = std::fs::remove_dir_all(&dir);
    }
}
